File: mouve/ViewPrototype/NodeConnectorView.py
from PySide.QtCore import *
from PySide.QtGui import *
from NodeTemporaryLinkView import *

class NodeConnectorView(QGraphicsWidget):
	def __init__(self, isOutput, parent=None):
		QGraphicsWidget.__init__(self, parent)

		self.isOutput = isOutput

		self._templink = None
		self._rect = NodeStyle.SocketSize
		self._pen = QPen(NodeStyle.SocketPen)

		gradient = QLinearGradient(0, self._rect.y(), 0, self._rect.height())
		gradient.setColorAt(0, NodeStyle.SocketGradientStart)
		gradient.setColorAt(1, NodeStyle.SocketGradientStop)
		self._brush = QBrush(gradient)		

		self._penWidth = 1.0 # !TODO: Use style
		self.setAcceptHoverEvents(True)

		self._animation = QPropertyAnimation(self, "penWidth")
		self._animation.setDuration(250)
		self._animation.setEasingCurve(QEasingCurve.InOutQuad);
		self._animation.setStartValue(self._penWidth)

	def paint(self, painter, option, widget):
		painter.setBrush(self._brush)
		painter.setPen(self._pen)
		painter.drawEllipse(self._rect)

	def boundingRect(self):
		return self._rect

	def hoverEnterEvent(self, event):
		self.setHighlight(True)

	def hoverLeaveEvent(self, event):
		self.setHighlight(False)

	def mousePressEvent(self, event):
		if event.button() == Qt.LeftButton:
			if self.isOutput:# !TODO: or self.isOutput == False:
				#self._templink = NodeTemporaryLinkView(self.mapToScene(self.centerPos()), event.scenePos(), None)
				#self.scene().addItem(self._templink)
				self._templink = NodeTemporaryLinkView(self.centerPos(), event.scenePos(), self)
				self.draggingLinkStarted.emit(self.socketView())

	def mouseMoveEvent(self, event):
		if self._templink is not None:
			self._templink.updateEndPosition(event.scenePos())

	def mouseReleaseEvent(self, event):
		if self._templink is not None:
			self.draggingLinkStopped.emit(self.socketView())
			itemColliding = self._canDrop(event.scenePos())
			if itemColliding is not None:
				self.draggingLinkDropped.emit(self.socketView(), itemColliding.socketView())
			self.scene().removeItem(self._templink)
			self._templink = None

	###############################################################################################

	def setHighlight(self, hightlight):
		if hightlight:
			self._animation.setStartValue(self._animation.currentValue())
			self._animation.setEndValue(2.0) # !TODO: Use style
			self._animation.start()
		else:
			self._animation.setStartValue(self._animation.currentValue())
			self._animation.setEndValue(1.0) # !TODO: Use style
			self._animation.start()

	def centerPos(self):
		return QPointF(self._rect.width() * 0.5, self._rect.height() * 0.5)

	def socketView(self):
		return self.parentObject()

	def _canDrop(self, scenePos):
		items = self.scene().items(scenePos, Qt.IntersectsItemShape, Qt.DescendingOrder)
		for item in items:
			# Did we dropped on another connector
			if item.type() == NodeConnectorView.Type and item != self:
				# Ensure we connect from input to output or the opposite
				if item.isOutput != self.isOutput:
					# Protect from constructing algebraic loops
					if item.socketView().nodeView() != self.socketView().nodeView():
						return item
		return None

	#
	# type property
	#

	Type = QGraphicsItem.UserType + 4

	def type(self):
		return NodeConnectorView.Type

	# penWidth property for mouse hover animation

	def getPenWidth(self):
		return self._penWidth

	def setPenWidth(self, val):
		self._penWidth = val
		self._pen.setWidthF(self._penWidth)
		self.update()

	penWidth = Property(float, getPenWidth, setPenWidth)

	# signals
	draggingLinkDropped = Signal(QGraphicsWidget, QGraphicsWidget)
	draggingLinkStarted = Signal(QGraphicsWidget)
	draggingLinkStopped = Signal(QGraphicsWidget)

File: mouve/NodeConnectorView.cpp
#include "NodeConnectorView.h"
#include "NodeStyle.h"
#include "NodeTemporaryLinkView.h"

#include <QGraphicsSceneMouseEvent>

NodeConnectorView::NodeConnectorView(bool isOutput, QGraphicsItem* parent)
	: QGraphicsWidget(parent, 0)
	, mIsOutput(isOutput)
	, mRect(NodeStyle::SocketSize)
	, mPen(NodeStyle::SocketPen)
	, mAnimation(this, "penWidth")
	, mTemporaryLink(nullptr)
{
	QLinearGradient gradient(0, mRect.y(), 0, mRect.height());
	gradient.setColorAt(0, NodeStyle::SocketGradientStart);
	gradient.setColorAt(1, NodeStyle::SocketGradientStop);
	mBrush = QBrush(gradient);
	mPen.setWidthF(1.0f); // !TODO: Use style

	setAcceptHoverEvents(true);

	mAnimation.setDuration(250);
	mAnimation.setEasingCurve(QEasingCurve::InOutQuad);
	mAnimation.setStartValue(penWidth());
}

void NodeConnectorView::setPenWidth(float penWidth)
{
	mPen.setWidthF(penWidth);
	update();
}

void NodeConnectorView::paint(QPainter* painter,
	const QStyleOptionGraphicsItem* option, QWidget* widget )
{
	painter->setBrush(mBrush);
	painter->setPen(mPen);
	painter->drawEllipse(mRect);
}

void NodeConnectorView::setHighlight(bool highlight)
{
	if(highlight)
	{
		mAnimation.setStartValue(mAnimation.currentValue());
		mAnimation.setEndValue(2.0); // !TODO: Use style
		mAnimation.start();
	}
	else
	{
		mAnimation.setStartValue(mAnimation.currentValue());
		mAnimation.setEndValue(1.0); // !TODO: Use style
		mAnimation.start();
	}
}

void NodeConnectorView::hoverEnterEvent(QGraphicsSceneHoverEvent* event)
{
	setHighlight(true);
}

void NodeConnectorView::hoverLeaveEvent(QGraphicsSceneHoverEvent* event)
{
	setHighlight(false);
}

void NodeConnectorView::mousePressEvent(QGraphicsSceneMouseEvent* event)
{
	if(event->button() == Qt::LeftButton)
	{
		if(isOutput()) // !TODO: or !isOutput()
		{ 
			mTemporaryLink = new NodeTemporaryLinkView
				(centerPos(), event->scenePos(), this);
			emit draggingLinkStarted(socketView());
		}
	}
}

void NodeConnectorView::mouseMoveEvent(QGraphicsSceneMouseEvent* event)
{
	if(mTemporaryLink != nullptr)
		mTemporaryLink->updateEndPosition(event->scenePos());
}

void NodeConnectorView::mouseReleaseEvent( QGraphicsSceneMouseEvent* event)
{
	if(mTemporaryLink != nullptr)
	{
		emit draggingLinkStopped(socketView());
		NodeConnectorView* itemColliding = canDrop(event->scenePos());
		if(itemColliding != nullptr)
			emit draggingLinkDropped(socketView(), itemColliding->socketView());
		scene()->removeItem(mTemporaryLink);
		delete mTemporaryLink;
		mTemporaryLink = nullptr;
	}
}

NodeConnectorView* NodeConnectorView::canDrop(const QPointF& scenePos)
{
	QList<QGraphicsItem*> items = 
		scene()->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder);
	foreach(QGraphicsItem* item, items)
	{
		// Did we dropped on another connector
		if(item->type() == NodeConnectorView::Type && item != this)
		{
			// Ensure we connect from input to output or the opposite
			NodeConnectorView* citem = static_cast<NodeConnectorView*>(item);
			if(citem->isOutput() != isOutput())
			{
				// Protect from constructing algebraic loops
				// !TODO: move to logic
				if(citem->socketView()->nodeView() != socketView()->nodeView())
				{
					return citem;
				}
			}
		}
	}
	return nullptr;
}


